src/ctx.rs

```diff
@@ -781,6 +781,35 @@ impl<'a> TryFromCtx<'a, usize> for &'a [u8] {
     }
 }
 
+impl<'a, Ctx: Copy, T: TryFromCtx<'a, Ctx, Error = error::Error>, const N: usize>
+    TryFromCtx<'a, Ctx> for [T; N]
+{
+    type Error = error::Error;
+    fn try_from_ctx(src: &'a [u8], ctx: Ctx) -> Result<(Self, usize), Self::Error> {
+        let mut offset = 0;
+        // SAFETY: This will only be read after everything was overwritten.
+        let mut buf: [T; N] = unsafe { zeroed() };
```

Comment:
this is not actually safe; `*element` line below will call the drop function on `T`, this is a typical pitfall of unsafe rust and using `mem::zeroed`, and hard to catch.

if this were e.g., a String, or some other complex object, it could theoretically segfault during it's drop routine

You have two options:
1. if you want to stick with unsafe, the proper solution is to use `mem::MaybeUninit`, with a write-only initialization of an array being the classic example/usecase.
2. if you don't want to stick with unsafe, i believe you could add a default bound to T, and then do, e.g.: `let mut buf: [T; N] = [T::default(); N]`

I'm ok with MaybeUninit approach, as it is less burdensome on the implementator to require default on their type, however we have to be extra careful during review that it is safe + sound :)

Reply:
I'm currently working on doing this with `MaybeUninit`, but I don't know how to turn `[MaybeUninit<T>;N]` into `[T;N]` due to the layout guarantees, transmute would be an option.

Reply:
Somewhat related, a fully safe impl that requires a SizeWith bound (e.g., no sizes depending on the ctx) which will rule out dynamically sized things like slices would look something like:
```rust
impl<
        'a,
        Ctx: Copy,
        T: TryFromCtx<'a, Ctx, Error = error::Error> + SizeWith<Ctx>,
        const N: usize,
    > TryFromCtx<'a, Ctx> for [T; N]
{
    type Error = error::Error;
    fn try_from_ctx(src: &'a [u8], ctx: Ctx) -> Result<(Self, usize), Self::Error> {
        let nbytes_required = N * T::size_with(&ctx);
        if nbytes_required > src.len() {
            panic!();
        }
        let buf: [T; N] =
            std::array::from_fn(|i| src.pread_with(i * T::size_with(&ctx), ctx).unwrap());

        Ok((buf, nbytes_required))
    }
}
```

I think the use of unsafe and `MaybeUninit` though is actually justified here, since your impl _is_ the "initializer"; again though we must be very careful during review we don't introduce unsoundness in this part of the library

Reply:
The issue is, that for one we are not handling errors properly, which will be hard, without using try_from_fn which is unstable and on the other hand can't reasonably introduce a Default bound to all types T.

Reply:
I'm currently at an implementation, which looks like this:
```rust
impl<'a, Ctx: Copy, T: TryFromCtx<'a, Ctx, Error = error::Error>, const N: usize>
    TryFromCtx<'a, Ctx> for [T; N]
{
    type Error = error::Error;
    fn try_from_ctx(src: &'a [u8], ctx: Ctx) -> Result<(Self, usize), Self::Error> {
        let mut offset = 0;

        // This is taken from the implementation of `MaybeUninit::uninit_array`, which is currently unstable.
        // SAFETY: An uninitialized `[MaybeUninit<_>; LEN]` is valid.
        let mut buf = unsafe { MaybeUninit::<[MaybeUninit<T>; N]>::uninit().assume_init() };

        for element in buf.iter_mut() {
            *element = MaybeUninit::new(src.gread_with::<T>(&mut offset, ctx)?);
        }

        Ok((buf.map(|element| unsafe { element.assume_init() }), offset))
    }
}
```
The array initialization was taken from `MaybeUninit::uninit_array`, which is unstable. This avoids having to introduce a `Copy` bound on T.

Reply:
This seems like the correct approach, although if we did add a copy bound I don’t think we need any unsafe at all.

One thing that does worry me is that if gread fails in that loop, any objects which got put in the array from earlier in the loop before failure, are never dropped. This is maybe ok but it does worry me a bit. There’s a way to fix this but it’s not great (reiterate back up to the failing index and assume_init it and that handles the drop as far as I know)

Reply:
Ah, i see what you mean now; yes we still need to `MaybeUninit` the array itself, which requires unsafe, if we have `Copy` bound; but we'd still have unsafe; in order to do full unsafe requires `Default`; i think what you've written is ok, though a `Copy` bound would ensure the type is "safer", but not quite enough.

Reply:
Here's another version with only 1 unsafe block, although there's a good chance the array is double-initialized, i.e., first from `from_fn` (likely a memset) then the pread loop, (would need to check assembly):
```rust
impl<'a, Ctx: Copy, T: TryFromCtx<'a, Ctx, Error = error::Error>, const N: usize>
    TryFromCtx<'a, Ctx> for [T; N]
{
    type Error = error::Error;
    fn try_from_ctx(src: &'a [u8], ctx: Ctx) -> Result<(Self, usize), Self::Error> {
        let mut offset = 0;
        use std::mem::MaybeUninit;

        let mut buf: [MaybeUninit<T>; N] = std::array::from_fn(|_| MaybeUninit::uninit());

        for element in buf.iter_mut() {
            *element = MaybeUninit::new(src.gread_with::<T>(&mut offset, ctx)?);
        }

        // SAFETY: we initialized each element above by preading them out, correctness
        // of the initialized element is guaranted by pread itself
        Ok((buf.map(|element| unsafe { element.assume_init() }), offset))
    }
}
```

i've also written a custom type with a custom drop impl and verified drop is only run once for each element (after the array leaves scope) whereas with the zeroed version, it is run twice (the `*element` deref as i mentioned)

Reply:
The array init looks pretty good to me, my biggest worry is `Drop` not being called, since `T` could potentially and very realistically contain heap allocated objects like `Box` or `Vec` which wouldn't be freed properly.

Reply:
And here is a version that drops the elements if we fail while preading; also for some reason i can't get transmute working, which should be a free cast, according to MaybeUninit docs, i think perhaps it doesn't work with const generics (yet)?

```rust
impl<'a, Ctx: Copy, T: TryFromCtx<'a, Ctx, Error = error::Error>, const N: usize>
    TryFromCtx<'a, Ctx> for [T; N]
{
    type Error = error::Error;
    fn try_from_ctx(src: &'a [u8], ctx: Ctx) -> Result<(Self, usize), Self::Error> {
        let mut offset = 0;
        use std::mem::MaybeUninit;

        let mut buf: [MaybeUninit<T>; N] = std::array::from_fn(|_| MaybeUninit::uninit());

        let mut error_ctx = None;
        for (idx, element) in buf.iter_mut().enumerate() {
            match src.gread_with::<T>(&mut offset, ctx) {
                Ok(val) => {
                    *element = MaybeUninit::new(val);
                }
                Err(e) => {
                    error_ctx = Some((e, idx));
                }
            }
        }
        if let Some((e, idx)) = error_ctx {
            for element in &mut buf[0..idx].iter_mut() {
                unsafe {
                    element.assume_init_drop();
                }
            }
            Err(e)
        } else {
            // SAFETY: we initialized each element above by preading them out, correctness
            // of the initialized element is guaranted by pread itself
            Ok((buf.map(|element| unsafe { element.assume_init() }), offset))
        }
    }
}
```

Reply:
I had trouble with `transmute` to, which is why I used `[T; N]::map` instead, but this map will with certainty be optimised out, since the function calls doesn't actually do anything. So far this looks great.

Reply:
The only thing that is missing, is that the first loop doesn't break on an error.

Reply:
I just added that and have this impl now:
```rust
impl<'a, Ctx: Copy, T: TryFromCtx<'a, Ctx, Error = error::Error>, const N: usize>
    TryFromCtx<'a, Ctx> for [T; N]
{
    type Error = error::Error;
    fn try_from_ctx(src: &'a [u8], ctx: Ctx) -> Result<(Self, usize), Self::Error> {
        let mut offset = 0;

        let mut buf: [MaybeUninit<T>; N] = core::array::from_fn(|_| MaybeUninit::uninit());

        let mut error_ctx = None;
        for (idx, element) in buf.iter_mut().enumerate() {
            match src.gread_with::<T>(&mut offset, ctx) {
                Ok(val) => {
                    *element = MaybeUninit::new(val);
                }
                Err(e) => {
                    error_ctx = Some((e, idx));
                    break;
                }
            }
        }
        if let Some((e, idx)) = error_ctx {
            for element in &mut buf[0..idx].iter_mut() {
                unsafe {
                    element.assume_init_drop();
                }
            }
            Err(e)
        } else {
            // SAFETY: we initialized each element above by preading them out, correctness
            // of the initialized element is guaranted by pread itself
            Ok((buf.map(|element| unsafe { element.assume_init() }), offset))
        }

    }
}
```

Reply:
yes push this; good catch on that `break` too, that would have been a serious error! :)

Reply:
I just pushed it, kudos for spotting the unsoundness.